### PR TITLE
Bring deep state to gluon-ts

### PR DESCRIPTION
**Pull request by rshyamsundar**
Implementation for the paper: https://papers.nips.cc/paper/8004-deep-state-space-models-for-time-series-forecasting

### CODE REVIEW

**Comment by szha** [comment]
Job PR-229/1 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-229/1/index.html

**Comment by lostella** [comment]
Looks like the following files are missing the required header:
```
File src/gluonts/model/deepstate/_estimator.py does not have correct header
File src/gluonts/model/deepstate/_network.py does not have correct header
File src/gluonts/model/deepstate/issm.py does not have correct header
File src/gluonts/model/deepstate/__init__.py does not have correct header
```

See the header here: https://github.com/awslabs/gluon-ts/blob/master/src/gluonts/transform.py

File: src/gluonts/time_feature/lag.py [hunk 1]
```diff
@@ -51,6 +51,28 @@ def get_granularity(freq_str: str) -> Tuple[int, str]:
...
+    "B": 365,  # yearly seasonality
+    "H": 168,  # weekly seasonality
+    "T": 1440,  # daily seasonality
+    "min": 1440,  # daily seasonality
+}
+
+
+def longest_period_from_frequency_str(freq_str: str) -> int:
+    multiple, granularity = get_granularity(freq_str)
+    return FREQ_LONGEST_PERIOD_DICT[granularity] // multiple
```
Comment (lostella) [inline src/gluonts/time_feature/lag.py]: This may be minor, but: this line will round down the period length. For example, with `freq_str="3W"` it will yield `52//3` which is 17, but to cover a year one would need to use 18.

File: src/gluonts/time_feature/lag.py [hunk 2]
```diff
@@ -51,6 +51,28 @@ def get_granularity(freq_str: str) -> Tuple[int, str]:
...
+
```
Comment (lostella) [inline src/gluonts/time_feature/lag.py]: Also: similar code to the one here appears in the `evaluation` package, [see here](https://github.com/awslabs/gluon-ts/blob/d6bc5468c28fa665b2d3d0e57e1a62221231f481/src/gluonts/evaluation/_base.py#L30-L54).

These two are of course different: in the `evaluation` package, for example, a seasonality of length 1 is used for `"D"` or `"W"`, and a seasonality of length 24 is used for `"H"`.

I'm not sure why that is the case, but I think that having similar pieces of code lying around can create confusion: I suggest keeping these definition here local to the `model.deepstate` package, since this is only used there. Then we will see about consolidating these.

**Comment by szha** [comment]
Job PR-229/2 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-229/2/index.html

File: src/gluonts/time_feature/lag.py (hunk 1 above)
Comment (lostella) [reply src/gluonts/time_feature/lag.py]: I mean, if a model looks over multiple periods, then this may be almost irrelevant; it's more relevant if one looks at a single period of course.

Comment (rshyamsundar) [reply src/gluonts/time_feature/lag.py]: yes, year has slightly more than 52 weeks. Since this is minor (only used to determine the training length) I went with 52 which is also the recommended frequency value in R forecast package.

File: src/gluonts/model/deepstate/_estimator.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        Flag to indicate whether to include trend component in the
+        state space model
+    past_length
+        This is the length of the training time series;
+        i.e., number of steps to unroll the RNN for before computing predictions.
+        Set this to (at most) the length of the shortest time series in the dataset.
+        (default: None, in which case the training length is set such that at least
+        `num_seasons_to_train` seasons are included in the training.
+        See `num_seasons_to_train`)
+    num_seasons_to_train
```
Comment (lostella) [inline src/gluonts/model/deepstate/_estimator.py]: Maybe a better name for this is `num_periods_to_train`? In my idea of naming things, a period is a full cycle of seasons

File: src/gluonts/time_feature/lag.py (hunk 1 above)
Comment (lostella) [reply src/gluonts/time_feature/lag.py]: I think 52 is fine, I was referring to `52//3` rounding down to something shorter than 52 weeks.

But again, I think this is minor, feel free to resolve this conversation.

File: src/gluonts/model/deepstate/_estimator.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+SEASON_INDICATORS_FIELD = "seasonal_indicators"
+
+
+class DeepStateEstimator(GluonEstimator):
+    """
+    Construct a DeepState estimator.
+    
+    This implements the deep state space model described in
+    [Rangapuram et. al. 2018]
```
Comment (benidis) [inline src/gluonts/model/deepstate/_estimator.py]: et al.

File: src/gluonts/model/deepstate/_estimator.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+    .. [Rangapuram et. al. 2018] Syama S. Rangapuram, Matthias W. Seeger,
```
Comment (benidis) [inline src/gluonts/model/deepstate/_estimator.py]: et al.

File: src/gluonts/model/deepstate/_estimator.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+
+    @validated()
+    def __init__(
+        self,
+        freq: str,
+        prediction_length: int,
+        add_trend: bool = False,
+        past_length: Optional[int] = None,
+        num_seasons_to_train: int = 4,
+        trainer: Trainer = Trainer(epochs=25, hybridize=False),
```
Comment (benidis) [inline src/gluonts/model/deepstate/_estimator.py]: Is there a reason to not use the default `Trainer()` like in the other models?

File: src/gluonts/model/deepstate/_estimator.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+        ), "The value of `prediction_length` should be > 0"
+        assert (
+            past_length is None or past_length > 0
+        ), "The value of `past_length` should be > 0"
+        assert num_layers > 0, "The value of `num_layers` should be > 0"
+        assert num_cells > 0, "The value of `num_cells` should be > 0"
+        assert (
+            num_eval_samples > 0
+        ), "The value of `num_eval_samples` should be > 0"
+        assert dropout_rate > 0, "The value of `dropout_rate` should be > 0"
```
Comment (benidis) [inline src/gluonts/model/deepstate/_estimator.py]: I think you should include equality here, i.e., `dropout_rate >= 0`

File: src/gluonts/model/deepstate/issm.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+        block_diagonal = _make_2_block_diagonal(
+            F=F, left=block_diagonal, right=blocks[i]
+        )
+
+    return block_diagonal
+
+
+def _make_2_block_diagonal(F, left: Tensor, right: Tensor) -> Tensor:
+    """
+    Creates a block diagonal matrix of shape (batch_size, m+n, m+n) where m and n are the sizes of
```
Comment (benidis) [inline src/gluonts/model/deepstate/issm.py]: fix documentation style

File: src/gluonts/model/deepstate/issm.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+    _block_diagonal = F.concat(
+        F.concat(left, zeros_off_diag, dim=3),
+        F.concat(zeros_off_diag_tr, right, dim=3),
+        dim=2,
+    )
+
+    return _block_diagonal
+
+
+class ISSM:
```
Comment (benidis) [inline src/gluonts/model/deepstate/issm.py]: would be nice to have some documentation here describing what the class does.

File: src/gluonts/distribution/lds.py [hunk 10]
```diff
@@ -315,7 +317,10 @@ def sample(
         # Sample the prior state.
         # samples_lat_state: (num_samples, batch_size, latent_dim, 1)
         state = MultivariateGaussian(
-            self.prior_mean, F.linalg_potrf(self.prior_cov)
+            self.prior_mean,
+            jitter_cholesky(
+                F, self.prior_cov, self.latent_dim, float_type=np.float32
+            ),
```
Comment (lostella) [inline src/gluonts/distribution/lds.py]: Could this be made optional? Like having `sampling_jitter=False` in the constructor, and set `sampling_jitter=True` when constructing the distribution in `DeepStateEstimator`?

File: src/gluonts/model/deepstate/_estimator.py [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+        use_feat_static_cat: bool = False,
+        cardinality: Optional[List[int]] = None,
+        embedding_dimension: int = 20,
+        issm: Optional[ISSM] = None,
+        scaling: bool = True,
+        time_features: Optional[List[TimeFeature]] = None,
+    ) -> None:
+        super().__init__(trainer=trainer)
+
+        assert (
```
Comment (jaheba) [inline src/gluonts/model/deepstate/_estimator.py]: You can use `prediction_length: pydantic.PositiveInt` and then remove the assert statements.

File: src/gluonts/model/deepstate/_estimator.py [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+            prediction_length > 0
+        ), "The value of `prediction_length` should be > 0"
+        assert (
+            past_length is None or past_length > 0
```
Comment (jaheba) [inline src/gluonts/model/deepstate/_estimator.py]: `past_length: pydantic.PositiveInt`

File: src/gluonts/model/deepstate/_estimator.py (hunk 6 above)
Comment (rshyamsundar) [reply src/gluonts/model/deepstate/_estimator.py]: yes..

* **epochs:** The total number of effective epochs by default is: `epochs * num_batches_per_epoch = 2500`. This is too much and wasteful for deep state model given that the structure of the time series is hard-coded (via SSM) and deep state is basically learning the parameters of this structure.
 
* **hybridize:** To avoid numerical issues when there is excess padding, I am using a jittering  method which does not work in symbolic mode.

File: src/gluonts/distribution/lds.py (hunk 10 above)
Comment (rshyamsundar) [reply src/gluonts/distribution/lds.py]: Ah.. forgot to add a comment about this. The `jitter_cholesky` method applies jitter only if the Cholesky decomposition fails for the given matrix and returns normal result without jitter if there is no problem. So this is essentially having the jitter only for problematic cases (btw, this happened only for the constant dataset in the evaluations).

I think `jitter_cholesky` was written this way to avoid having further option/flag like `sampling_jitter` on the caller side.

File: src/gluonts/model/deepstate/issm.py (hunk 8 above)
Comment (rshyamsundar) [reply src/gluonts/model/deepstate/issm.py]: Thanks

File: src/gluonts/model/deepstate/_estimator.py (hunk 7 above)
Comment (rshyamsundar) [reply src/gluonts/model/deepstate/_estimator.py]: Right. Thanks.

File: src/gluonts/model/deepstate/_estimator.py (hunk 12 above)
Comment (rshyamsundar) [reply src/gluonts/model/deepstate/_estimator.py]: I think we had a problem with this earlier so we used asserts in other estimators as well. So to be consistent with other estimators I will leave it like this for now.

File: src/gluonts/model/deepstate/issm.py (hunk 9 above)
Comment (rshyamsundar) [reply src/gluonts/model/deepstate/issm.py]: Added, thanks.

File: src/gluonts/model/deepstate/_estimator.py (hunk 12 above)
Comment (jaheba) [reply src/gluonts/model/deepstate/_estimator.py]: Right, I think we should address this at a different time.

Do you know what the problems where?

File: src/gluonts/distribution/lds.py (hunk 10 above)
Comment (dcmaddix) [reply src/gluonts/distribution/lds.py]: Yes exactly. It only applies the jitter when it is needed and otherwise does the standard Cholesky decomposition.

**Comment by szha** [comment]
Job PR-229/6 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-229/6/index.html

**Comment by lixixibj** [comment]
> Job PR-229/1 is complete.
> Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-229/1/index.html

Hi, the URL of the docs cannot be accessed. Please give me a valid URL.

**Comment by lixixibj** [comment]
Hi, in the paper deep state space, the emission_coeff transition_coeff is time-varying, however, in the code in gluon-ts, these coefficients seem to be time-invariant and fixed. I don't know which actually is true. Could you please update the code of the paper?

**Comment by rshyamsundar** [comment]
For seasonal models the emission coefficient is time/season dependent and other are constants because of the way we implement the innovation state space models.

Check the Appendix A.1 in the [long version](https://papers.nips.cc/paper/8004-deep-state-space-models-for-time-series-forecasting-supplemental.zip) to see the specific instantiations (that we used in the experiments) of the more general state space model discussed in the paper.
